Add AutoTLS client spec (#682)
libp2p/specs: tls/autotls-client.md

## General Flow
The following is the general flow of a successful certificate request and subsequent issuance using AutoTLS.
Here, "node" refers to the machine running a libp2p peer and requesting the challenge,
while "broker" and "AutoTLS broker", which are used interchangeably, refer to the server that will fulfil the ACME challenge on behalf of the node.

1. Node requests a challenge from the ACME server.
2. Node sends the challenge to the broker.
3. Broker tests node and sets DNS record (fulfilling challenge).
4. Node queries DNS until it sees that the broker has fulfilled the challenge.
5. Node signals to ACME server that challenge is fulfilled.
6. ACME server checks challenge in broker.
7. Node sends CSR to finalize certificate request.
8. Node polls ACME server until certificate is ready for download.
9. Node downloads certificate.

## Parameters

| Parameter                | Description                                                      | Reasonable Default |
|--------------------------|------------------------------------------------------------------|--------------|

> | `max_dns_retries` | The maximum number of DNS queries that the node SHOULD make before giving up | ???  |

Reviewer: `p2p-forge/client` uses time-based timeout (3 minutes) with exponential backoff rather than fixed retry counts

> | `max_dns_timeout` | The maximum number of seconds a node SHOULD wait for DNS records to be set | ???  |

Reviewer: iirc `p2p-forge/client` gives up waiting for TXT record after 3min I think -- at this point something went wrong and the entire registration needs to be resumed

> | `max_acme_poll_retries` | The maximum number of GET requests that the node SHOULD issue to ACME server before giving up | ???  |

Reviewer: ACME specs recommend polling until timeout rather than fixed retry counts

> | `max_acme_timeout` | The maximum number of seconds a node SHOULD wait for an ACME resource status to change | ???  |

Reviewer: Let's Encrypt recommendation for challenge completion is ~10minutes, p2p-forge client aborts sooner, around 3 minutes I think

	3. Node generates `sig`, `headers` and `payload` as follows, where `peer-privkey` is the private key of the node's libp2p peer and `multiaddrs` is a list of string representations of the libp2p peer's multiaddresses:
	```
		sig = base64URL(
			peer-privkey.sign(
				bytes(varint + "challenge-node={challenge-node}") +
				bytes(varint + "hostname={hostname}") +
				bytes(varint + "server-public-key={public-key}")
			)
		)

		headers = {
			"Content-Type": "application/json",
			"User-Agent": "some-user-agent",
			"authorization": "libp2p-PeerID public-key=\"{nodePublicKeyB64}\", opaque=\"{opaque}\", challenge-server=\"{challengeServer}\", sig=\"{sig}\""
		}

        payload = {
            "value": keyAuthorization,
            "addresses": multiaddrs
        }
	```

    **Note:** `varint` is a protobuf [varint](https://protobuf.dev/programming-guides/encoding/#varints) field that encodes the length of each of the `key=value` string.

>     **Note:** The node SHOULD include only multiaddresses containing public IPv4 addresses in `multiaddrs`.

Reviewer: iirc there is more this (IPv6, filtering):

- The node MUST include only publicly reachable addresses
- For IPv4: exclude private ranges
- For IPv6: exclude private ranges and NAT64 translated addresses
- Exclude relay addresses (containing `/p2p-circuit`)

	4. Node sends a POST request to `/v1/_acme-challenge` endpoint using `payload` as HTTP body and `headers` as HTTP headers.
	5. Node SHOULD save the `bearer` token from the `authentication-info` response header, and use it for following requests to the AutoTLS broker.

## Signalling challenge completion to ACME server

> 1. Node SHOULD query DNS records (`TXT _acme-challenge.{b36peerid}.libp2p.direct` and `A dashed-public-ip-address.{b36peerid}.libp2p.direct`) until they are set by the AutoTLS broker.

Reviewer: no point in querying `A` if `TXT` is not published by the registration endpoint:

```suggestion
1. Node SHOULD query DNS records (`TXT _acme-challenge.{b36peerid}.libp2p.direct`) until they are set by the AutoTLS broker using exponential backoff (double the interval after each failed attempt).
```

**Note:** Here, `dashed-public-ip-address` is the public IPv4 address on in which the node received the confirmation dial from the broker.

> For example, if the node has two public IPv4 addresses `1.1.1.1` and `8.8.8.8`, and the broker dialed it through `1.1.1.1`, then the node SHOULD query the `A 1-1-1-1.{b36peerid}.libp2p.direct`.

Reviewer: ```suggestion
For example, if the node has two public IPv4 addresses `1.1.1.1` and `8.8.8.8`, and the broker dialed it through `1.1.1.1`, then the node SHOULD query the `A 1-1-1-1.{b36peerid}.libp2p.direct`.
For IPv6: Replace colons with dashes, handle leading/trailing colons as zeros (`2001:db8::1` → `2001-db8--1`, `2001:db8::` → `2001-db8--0`), and query `AAAA 2001-db8--1.{b36peerid}.libp2p.direct`.
```

**Note:** The node SHOULD NOT send more than `max_dns_retries` DNS requests.
After `max_dns_timeout`, the communication is considered failed.

> What to do after `max_dns_timeout` has passed is left as an implementation decision.

Reviewer: nit: imo the spec should clearly say that at this point the certificate request flow SHOULD be aborted and retried later.
 
(Do not attempt to ask ACME for cert if DNS TXT record is not confirmed to exist – we don't want to hammer Let's Encrypt with obviously failing requests)

**Note:** The node SHOULD NOT send more than `max_acme_poll_retries` poll requests to the ACME server.
After `max_acme_timeout`, the communication has failed.
What to do after `max_acme_timeout` has passed is left as an implementation decision.

## Downloading certificate
1. Node finalizes the certificate request:
    1. Generate CSR for the `*.{b36peerid}.libp2p.direct` domain.
    2. Encode the CSR with URL safe base 64 (`b64CSR`).
    3. Send a `kid` signed POST request to `finalizeUrl` with JSON HTTP body of `{"csr": b64CSR}`.
2. Node MUST poll ACME server by sending GET requests to `orderUrl` until the ACME server's response contains a `status` field with a value different than `processing`.

**Note:** The node SHOULD NOT send more than `max_acme_poll_retries` poll requests to the ACME server.
After `max_acme_timeout`, the communication has failed.

> What to do after `max_acme_timeout` has passed is left as an implementation decision.

Reviewer: Similar ask: spec should clearly note that there should be exponential backoff for retry attempts.

4. From the ACME server's response, node saves `orderUrl`, `chalUrl` and `finalizeUrl`:
    ```
    orderUrl: https://acme-staging-v02.api.letsencrypt.org/acme/order/201427344/24815752984
    chalUrl: https://acme-staging-v02.api.letsencrypt.org/acme/chall/201427344/17523856954/k8-vYA
    finalizeUrl: https://acme-staging-v02.api.letsencrypt.org/acme/finalize/201427344/24815752984
    ```

5. Node generates `keyAuthorization` (`jP5hwrZwCbP_qeeET_qAa9pgG0YulNaR0ivruESzCrE`) from the following `dns-01` object:
    ```json
    {
      "type": "dns-01",
      "url": "https://acme-staging-v02.api.letsencrypt.org/acme/chall/201427344/17523856954/k8-vYA",
      "status": "pending",
      "token": "nE2YGvFXzAy6UsFjYxqYOyA6rxZ7VJeQppsQ72hHyPM"
    }
    ```

6. Node authenticates with AutoTLS broker at `"https://registration.libp2p.direct/v1/_acme-challenge"` (refer to the [peer ID authentication spec](https://github.com/libp2p/specs/blob/master/http/peer-id-auth.md) for guidance and examples) and sends the following JSON body:
    ```json
    {
      "value": "jP5hwrZwCbP_qeeET_qAa9pgG0YulNaR0ivruESzCrE",
      "addresses": [
        "/ip4/142.93.194.175/tcp/49309"
      ]
    }
    ```

> 

Reviewer: ```suggestion

**User-Agent Header**: Implementations SHOULD include a descriptive User-Agent header that identifies the client library name and version or the host application (e.g., "kubo/0.17.0", "helia/2.0.0"). 

```
